0015-automatic-cluster-upgrades/0015-automatic-cluster-upgrades.md

```diff
@@ -0,0 +1,95 @@
+# RFC 0015 - Automatic Cluster Upgrades
+
+## User stories
+
+- As a user, and also as KaaS team provider, I would like to have the clusters running in the latest version automatically, and not worry about CVEs or EOL of the cluster components.
+
+- As a user, I would like to define maintenance windows to ensure clusters are not upgraded outside of these hours.
+
+- As a user I would like to mark a cluster as frozen and block upgrades for certain period of time.
+
+- As a user I want to define a policy that enforces some environments are upgraded before than others to an agreed version.
+
+- As a user I want to schedule a upgrade for a certain day/time.
+
+- As a user, and also as KaaS team provider, I want to capture the result of upgrade.
+
+## Reasoning
+
+Our customers need to plan every upgrade on their side and manage the upgrades for every GS version. Every customer has their own requirement when they need to schedule upgrades (freeze periods, maintenance windows,...). At the same time, Cloud Native projects and Giant Swarm are releasing more versions every time. It provokes a lot of time preparing and executing these upgrades which is not usually given much value to any of the stakeholders. 
+
+Giant Swarm designed a versioning schema (semver), which includes all components on the stack, to ensure idempotency and expose the level of impact that changes bring to the customer. Patches and Minor releases should have no impact on the customer workloads, which mean can be applied in the environments at any time. 
+
+## Current state
+
+Right now the automatic upgrades are carried by Solution Engineers agreeing with the customer on upgrade windows and environments selection. The idea is to automate all this behavior and information under a Kubernetes operator(s) and Custom Resource(s).
+
+The Solution engineer plans cluster upgrades based on cluster type (environment, team,...) and maintenance windows offered by the customers. The upgrade is just a change on the cluster version field on the cluster Custom Resource (CR).
+
+Based on the fact Giant Swarm is moving to Cluster API this is story let us define a new greenfield scenario where we can design a model that can be more generic and not focus only in Giant Swarm platform. The idea would be contribute as much as possible with upstream due the fact most of this functionality would be useful for everyone that manages clusters.
+
+## Concepts
+
+In cluster upgrade hive we have discuss around entities and ideas how this would look like.
+
+### Cluster Upgrade entity
+
+There are different reason why we are interested in having a object to define the upgrade intent. 
+
+- As we define in the user stories, we/customers would like to schedule upgrades on certain date from running version to a specific one. That way customer can plan upgrades ahead and so do we.
+
+- We have a registry of upgrades done to a cluster for auditing and tracking purposes.
```

Comment:
Let's work out how this could work. Say I have a `Cluster` CR with `name: foo01` and an associated KCP with `version: 1.18.16`.

I then create a hypothetical `ClusterUpgrade` with a reference to the `foo01` instance of `Cluster` and a `version: 1.19.8` to express the desire to upgrade.

An _upgrade-operator_ then processes that `ClusterUpgrade`, orchestrates the upgrades where needed, sets the `status` on the `ClusterUpgrade` and we're done. But if we want to keep that `ClusterUpgrade` resource around as a record of upgrade history, how do we make sure it's not processed again?

A simple answer could be adding a label like `upgrade-operator.giantswarm.io/upgrade-done: true` or something, but is that going to be performant enough?

The average number of clusters per installation is probably below 10, so not too bad. What about how many upgrades do clusters go through in their lifetime, on average?

Reply:
That's a good question, that should be enough to have a cleanup for older resources.

I'm not really sure if metrics are the right thing, but let's first narrow down what information about upgrades we care about keeping:

- that an upgrade happened: for sure 😉 
- how it was triggered: auto / manual
- outcome: success / fail (rollback?)
- from GS release?
- to GS release
- do we want to have a list of all components that were actually changed? like upgrade from `v13.0.0` to `v13.0.1` would only have changed the version of `kvm-operator`, do we need that level of detail?
- anything else?

Reply:
> do we want to have a list of all components that were actually changed? like upgrade from v13.0.0 to v13.0.1 would only have changed the version of kvm-operator, do we need that level of detail?

I guess knowing from which GS version we started and which we ended, will give us all info

> anything else?

When it was triggered and when it was done.

Reply:
From the chat we had:

- storing that information in a metric doesn't really fit the concept, with `from` and `to` versions which would have to be stored as labels you'd have unbounded sets of values and to represent start time and end time you'd need to do something like emitting the metric with value 1 when an upgrade is in progress and value 0 when it's not and then rely on history in cortex/grafana cloud or something
- even a couple of hundred of objects on large installations shouldn't be an issue, in some cases we're already dealing with thousands of objects and things work fine

TL;DR I'm happy enough with this approach.

I've also taken the list of bits of information we want to preserve and added to suggestion of how the status of `ClusterUpgrade` could look like.